# Introduction
Data preparation of the TechDebt dataset. Concretely, from the following tables:
- GIT_COMMITS
- GIT_COMMITS_CHANGES
- JIRA_ISSUES
- SONAR_ANALYSIS
- SONAR_ISSUES
- SONAR_MEASURES

## Library Packages

In [74]:
# Import libraries and packages
# Miscellaneous libraries
import numpy as np
import pandas as pd
import os
from datetime import datetime
import collections
from prettytable import PrettyTable

## Data Preparation

#### Define the path of the files

In [75]:
# Define the path of the data files
path = '../data/raw/'
path_git_commits = path + 'GIT_COMMITS.csv'
path_git_commits_changes = path + 'GIT_COMMITS_CHANGES.csv'
path_jira_issues = path + 'JIRA_ISSUES.csv'
path_sonar_analysis = path + 'SONAR_ANALYSIS.csv'
path_sonar_issues = path + 'SONAR_ISSUES.csv'
path_sonar_measures = path + 'SONAR_MEASURES.csv'

# Ensure the input file exist
assert os.path.isfile(path_git_commits), f'{path_git_commits} not found. Is it a file?'
assert os.path.isfile(path_git_commits_changes), f'{path_git_commits_changes} not found. Is it a file?'
assert os.path.isfile(path_jira_issues), f'{path_jira_issues} not found. Is it a file?'
assert os.path.isfile(path_sonar_analysis), f'{path_sonar_analysis} not found. Is it a file?'
assert os.path.isfile(path_sonar_issues), f'{path_sonar_issues} not found. Is it a file?'
assert os.path.isfile(path_sonar_measures), f'{path_sonar_measures} not found. Is it a file?'

#### Read the files

In [76]:
# Read the files
#git_commits_changes = pd.read_csv(path_git_commits_changes)
#git_commits = pd.read_csv(path_git_commits)
jira_issues = pd.read_csv(path_jira_issues)
sonar_analysis = pd.read_csv(path_sonar_analysis)
sonar_issues = pd.read_csv(path_sonar_issues)
sonar_measures = pd.read_csv(path_sonar_measures)

#### Define selected variables
In the following section we are only selecting the useful variables for the project. The election process has been studied previusly, in the Data Understanding step.

In [77]:
# Define variables of interest for each dataframe
#git_commits_changes_names = ['COMMIT_HASH','DATE','LINES_ADDED','LINES_REMOVED']
#git_commits_names = ['PROJECT_ID','COMMIT_HASH','AUTHOR','AUTHOR_DATE','AUTHOR_TIMEZONE']
jira_issues_names = ['HASH']
sonar_analysis_names = ['PROJECT_ID','ANALYSIS_KEY','REVISION']
sonar_issues_names = ['CREATION_ANALYSIS_KEY','SEVERITY','STATUS','EFFORT','MESSAGE','START_LINE','END_LINE','CLOSE_ANALYSIS_KEY']
sonar_measures_names = ['analysis_key','complexity' ,'cognitive_complexity', 'coverage', 'duplicated_blocks', 'duplicated_files', 
                        'duplicated_lines_density', 'violations','blocker_violations','critical_violations','major_violations','minor_violations','info_violations','false_positive_issues','open_issues','reopened_issues','confirmed_issues', 'sqale_debt_ratio','code_smells','bugs','reliability_rating','vulnerabilities','security_rating','files', 'comment_lines_density']

In [78]:
# Select variables of interest
#git_commits = git_commits[git_commits_names]
jira_issues = jira_issues[jira_issues_names]
sonar_analysis = sonar_analysis[sonar_analysis_names]
sonar_issues = sonar_issues[sonar_issues_names]
sonar_measures = sonar_measures[sonar_measures_names]

In [79]:
# Select columns of interest
dtypes = ['uint8','int16', 'int32', 'int64', 'float16', 'float32', 'float64', 'object']

## Global functions

In [80]:
def delete_na(name_tables, dataframes, dtypes):
    '''
    Objective:
        - Delete all NA's from the dataframe passed
    Input:
        - Name_tables : String of the names of the tables
        - Dataframe : String of the tables and their selected columns
        - Numerical : Numerical types
        
    Output: 
        - Table with the dataset sizes before and after deleting all NA:
        ["Table name", "Total number of rows", "Number of NA", "Total number of rows after deleting NA"]
    '''
    table = PrettyTable()
    table.field_names = ["Table name", "Total number of rows", "Number of NA", "Total number of rows after deleting NA"]
    for i in range(len(name_tables)):
        dataframe_numerical = dataframes[i].select_dtypes(include=dtypes)
        total_rows = dataframe_numerical.shape[0]
        # Rows with na's
        rows_with_NA = sum([True for idx,row in dataframe_numerical.iterrows() if any(row.isnull())])
        # Delete rows that contain na's
        dataframe_numerical = dataframe_numerical.dropna()
        total_rows_without_NA = dataframe_numerical.shape[0]
        table.add_row([name_tables[i] , total_rows, rows_with_NA,total_rows_without_NA])
        dataframes[i] = dataframe_numerical
    print(table)
    return dataframes


In [81]:
def analyse_categorical_variables(table_names, variable_names, dataframes):
    '''
    Objective:
        - Analyse the categorical variables to be encoded
    Input:
        - Table_names : String of the names of the tables
        - Variable_names : String of the categorical variables corresponding to the table
        - Dataframes : String of the tables and their selected columns
        
    Output: 
        - Table with the categorical variables levels
        ["Table name", "Variable name", "Number of levels", "Types"]
    '''
    table = PrettyTable()
    table.field_names = ["Table name", "Variable name", "Number of levels", "Types"]
    for i in range(len(table_names)):
        for j in range(len(variable_names)):
            table.add_row([table_names[i], variable_names[j], len(dataframes[i][variable_names[j]].unique()), dataframes[i][variable_names[j]].unique()])
    print(table)    

In [82]:
def one_hot_encoding(table, variable):
    variable_dummies = pd.get_dummies(table[variable])
    table = table.drop(variable, axis=1)
    table = table.join(variable_dummies)
    return table
    

### NA values
Deleting all NA values from the tables by using the global function implemented above delete_na().

In [83]:
table_names = ["SONAR_MEASURES", "SONAR_ISSUES", "SONAR_ANALYSIS", "JIRA_ISSUES"]
tables = [sonar_measures, sonar_issues, sonar_analysis, jira_issues]
[sonar_measures, sonar_issues, sonar_analysis, jira_issues] = delete_na(table_names, tables, dtypes)

+----------------+----------------------+--------------+----------------------------------------+
|   Table name   | Total number of rows | Number of NA | Total number of rows after deleting NA |
+----------------+----------------------+--------------+----------------------------------------+
| SONAR_MEASURES |        66711         |      34      |                 66677                  |
|  SONAR_ISSUES  |       1024614        |    296205    |                 728409                 |
| SONAR_ANALYSIS |        67550         |     839      |                 66711                  |
|  JIRA_ISSUES   |        22007         |      0       |                 22007                  |
+----------------+----------------------+--------------+----------------------------------------+


### Categorical Values
The next step is to analyse the categorical variables and encoding them.
For the SONAR_MEASURES, JIRA_ISSUES, SONAR_ANALYSIS table (add more if necessary) there are not categorical varibles.

In [84]:
table_names = ["SONAR_ISSUES"]
variable_names = ["SEVERITY", "STATUS", "EFFORT"]
dataframes = [sonar_issues]
analyse_categorical_variables(table_names, variable_names, dataframes)


+--------------+---------------+------------------+--------------------------------------------------------------------+
|  Table name  | Variable name | Number of levels |                               Types                                |
+--------------+---------------+------------------+--------------------------------------------------------------------+
| SONAR_ISSUES |    SEVERITY   |        5         |           ['INFO' 'MINOR' 'MAJOR' 'CRITICAL' 'BLOCKER']            |
| SONAR_ISSUES |     STATUS    |        1         |                             ['CLOSED']                             |
| SONAR_ISSUES |     EFFORT    |       181        | [1.0000e+01 2.0000e+00 9.0000e+01 1.0000e+00 5.0000e+00 8.0000e+00 |
|              |               |                  |  1.2000e+01 1.4000e+01 3.0000e+01 2.0000e+01 6.0000e+01 2.5000e+01 |
|              |               |                  |  1.1000e+01 1.9000e+01 2.8000e+01 5.6000e+01 6.0000e+00 1.3000e+01 |
|              |               |

As can be seen in the chunk above, the SEVERITY and STATUS variables have 5 and 1 levels respectively. In our case, we have performed the One-hot encoding for the SEVERITY variable. For the STATUS variable, efore deleting all NA, there was the OPENED level. However, all rows with an OPENED status contained NA, which means that for this variable we only have the CLOSED level. 
When joining the tables, we will calulate the mean of each types each author has.

In [85]:
sonar_issues = one_hot_encoding(sonar_issues, "SEVERITY")

In [86]:
sonar_issues = one_hot_encoding(sonar_issues, "STATUS")
sonar_issues.head()

CREATION_ANALYSIS_KEY  EFFORT  \
41   AWd5_psxC4KKKThc-qK6    10.0   
198  AWd5_psxC4KKKThc-qK6    10.0   
199  AWd5_psxC4KKKThc-qK6    10.0   
200  AWd5_psxC4KKKThc-qK6    10.0   
201  AWd5_psxC4KKKThc-qK6    10.0   

                                               MESSAGE  START_LINE  END_LINE  \
41   Do not forget to remove this deprecated code s...       104.0     104.0   
198  Do not forget to remove this deprecated code s...       354.0     354.0   
199  Do not forget to remove this deprecated code s...       600.0     600.0   
200  Do not forget to remove this deprecated code s...       609.0     609.0   
201  Do not forget to remove this deprecated code s...       619.0     619.0   

       CLOSE_ANALYSIS_KEY  BLOCKER  CRITICAL  INFO  MAJOR  MINOR  CLOSED  
41   AWeDrnWEC4KKKThcAtWf        0         0     1      0      0       1  
198  AWeMr1K3C4KKKThcB9hi        0         0     1      0      0       1  
199  AWeMr1K3C4KKKThcB9hi        0         0     1      0      0       1  
200  AWeMr1K3C4KKKThcB9hi        0         0     1      0      0       1  
201  AWeMr1K3C4KKKThcB9hi        0         0     1      0      0       1

### MESSAGE variable
In the following section, we will encode the MESSAGE variable for the SONAR_ISSUES table. For that variable, we will calulate the length fo the message for each issue, and reassigning the column with that new value instead of the initial message.

In [87]:
message_length = []
for row in sonar_issues['MESSAGE']:
    message_length.append(len(row))
message_length
len(message_length)

728409

In [88]:
# reassign the MESSAGE variable to its length instead of the initial string 
sonar_issues['MESSAGE'] = message_length
sonar_issues.head()

CREATION_ANALYSIS_KEY  EFFORT  MESSAGE  START_LINE  END_LINE  \
41   AWd5_psxC4KKKThc-qK6    10.0       53       104.0     104.0   
198  AWd5_psxC4KKKThc-qK6    10.0       53       354.0     354.0   
199  AWd5_psxC4KKKThc-qK6    10.0       53       600.0     600.0   
200  AWd5_psxC4KKKThc-qK6    10.0       53       609.0     609.0   
201  AWd5_psxC4KKKThc-qK6    10.0       53       619.0     619.0   

       CLOSE_ANALYSIS_KEY  BLOCKER  CRITICAL  INFO  MAJOR  MINOR  CLOSED  
41   AWeDrnWEC4KKKThcAtWf        0         0     1      0      0       1  
198  AWeMr1K3C4KKKThcB9hi        0         0     1      0      0       1  
199  AWeMr1K3C4KKKThcB9hi        0         0     1      0      0       1  
200  AWeMr1K3C4KKKThcB9hi        0         0     1      0      0       1  
201  AWeMr1K3C4KKKThcB9hi        0         0     1      0      0       1

NUMBER_OF_LINES variable

In the following cells we will proceed to computate the length mean per issue with the START_LINE and END_LINE variables.

In [89]:
issue_length = []
for index, row in sonar_issues.iterrows():
    diff = row['END_LINE'] - row['START_LINE']
    issue_length.append(diff)
len(issue_length)

728409

In [90]:
unique_values = set(issue_length)
len(unique_values)


89

In [91]:
print(unique_values)

{0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 27.0, 28.0, 31.0, 32.0, 33.0, 35.0, 37.0, 38.0, 39.0, 40.0, 41.0, 42.0, 43.0, 44.0, 45.0, 46.0, 49.0, 51.0, 52.0, 54.0, 57.0, 60.0, 61.0, 65.0, 66.0, 69.0, 70.0, 71.0, 73.0, 74.0, 78.0, 81.0, 82.0, 84.0, 90.0, 105.0, 109.0, 115.0, 116.0, 123.0, 130.0, 133.0, 134.0, 137.0, 168.0, 171.0, 186.0, 188.0, 189.0, 197.0, 210.0, 214.0, 215.0, 249.0, 251.0, 311.0, 321.0, 330.0, 334.0, 401.0, 409.0, 426.0}


In [92]:
sonar_issues = sonar_issues.drop('START_LINE', axis=1)
sonar_issues = sonar_issues.drop('END_LINE', axis=1)
sonar_issues['ISSUE_CODE_LENGTH'] = issue_length
sonar_issues.head()

CREATION_ANALYSIS_KEY  EFFORT  MESSAGE    CLOSE_ANALYSIS_KEY  BLOCKER  \
41   AWd5_psxC4KKKThc-qK6    10.0       53  AWeDrnWEC4KKKThcAtWf        0   
198  AWd5_psxC4KKKThc-qK6    10.0       53  AWeMr1K3C4KKKThcB9hi        0   
199  AWd5_psxC4KKKThc-qK6    10.0       53  AWeMr1K3C4KKKThcB9hi        0   
200  AWd5_psxC4KKKThc-qK6    10.0       53  AWeMr1K3C4KKKThcB9hi        0   
201  AWd5_psxC4KKKThc-qK6    10.0       53  AWeMr1K3C4KKKThcB9hi        0   

     CRITICAL  INFO  MAJOR  MINOR  CLOSED  ISSUE_CODE_LENGTH  
41          0     1      0      0       1                0.0  
198         0     1      0      0       1                0.0  
199         0     1      0      0       1                0.0  
200         0     1      0      0       1                0.0  
201         0     1      0      0       1                0.0

## Joints

Resum:
- 66711 rows in sonar analysis
- 719186 rows un cop fent el join amb la variable creation_Analysis key
- 728409 rows un cop fent el join amb el CLOSE_ANALYSIS_KEY
- 1447595 rows dels dos joints 
- 1200310 unique rows en total dels dos joints
- 1197656 rows un cop fent el join del SONRA_MEASURES i SONAR_COMPLETE (sonar_analysis + sonar_issues)
- 1083655 unique rows en total dels dos joints

(ho he calulat abans amb datafrme.shape[0] però els chuncks els he eliminat per netejar el codi)

In [93]:
# Joining SONAR_ANALYSIS with SONAR_ISSUES
sonar_complete_1 = pd.merge(sonar_issues, sonar_analysis, left_on='CREATION_ANALYSIS_KEY', right_on='ANALYSIS_KEY', how='inner')
sonar_complete_2 = pd.merge(sonar_issues, sonar_analysis, left_on='CLOSE_ANALYSIS_KEY', right_on='ANALYSIS_KEY', how='inner')
sonar_complete_1 = sonar_complete_1.drop('CREATION_ANALYSIS_KEY', axis=1)
sonar_complete_1 = sonar_complete_1.drop('CLOSE_ANALYSIS_KEY', axis=1)
sonar_complete_2 = sonar_complete_2.drop('CREATION_ANALYSIS_KEY', axis=1)
sonar_complete_2 = sonar_complete_2.drop('CLOSE_ANALYSIS_KEY', axis=1)

sonar_complete = pd.concat([sonar_complete_1, sonar_complete_2])

In [94]:
sonar_complete.shape[0]

1447595

In [95]:
# deleting duplicated rows
sonar_complete = sonar_complete[sonar_complete.duplicated()]

In [96]:
sonar_complete.shape[0]

1200310

In [97]:
sonar_complete.head()

EFFORT  MESSAGE  BLOCKER  CRITICAL  INFO  MAJOR  MINOR  CLOSED  \
1    10.0       53        0         0     1      0      0       1   
2    10.0       53        0         0     1      0      0       1   
3    10.0       53        0         0     1      0      0       1   
4    10.0       53        0         0     1      0      0       1   
5    10.0       53        0         0     1      0      0       1   

   ISSUE_CODE_LENGTH          PROJECT_ID          ANALYSIS_KEY  \
1                0.0  org.apache:cayenne  AWd5_psxC4KKKThc-qK6   
2                0.0  org.apache:cayenne  AWd5_psxC4KKKThc-qK6   
3                0.0  org.apache:cayenne  AWd5_psxC4KKKThc-qK6   
4                0.0  org.apache:cayenne  AWd5_psxC4KKKThc-qK6   
5                0.0  org.apache:cayenne  AWd5_psxC4KKKThc-qK6   

                                   REVISION  
1  fc186f50392bf8c804b1d314971c92efa5f52fc3  
2  fc186f50392bf8c804b1d314971c92efa5f52fc3  
3  fc186f50392bf8c804b1d314971c92efa5f52fc3  
4  fc186f50392bf8c804b1d314971c92efa5f52fc3  
5  fc186f50392bf8c804b1d314971c92efa5f52fc3

In [98]:
# Joining SONAR_ANALYSIS with SONAR_MEASURES
sonar_complete = pd.merge(sonar_complete, sonar_measures, left_on='ANALYSIS_KEY', right_on='analysis_key', how='inner')
sonar_complete = sonar_complete.drop('ANALYSIS_KEY', axis=1)
sonar_complete.head()

EFFORT  MESSAGE  BLOCKER  CRITICAL  INFO  MAJOR  MINOR  CLOSED  \
0    10.0       53        0         0     1      0      0       1   
1    10.0       53        0         0     1      0      0       1   
2    10.0       53        0         0     1      0      0       1   
3    10.0       53        0         0     1      0      0       1   
4    10.0       53        0         0     1      0      0       1   

   ISSUE_CODE_LENGTH          PROJECT_ID  ... reopened_issues  \
0                0.0  org.apache:cayenne  ...               0   
1                0.0  org.apache:cayenne  ...               0   
2                0.0  org.apache:cayenne  ...               0   
3                0.0  org.apache:cayenne  ...               0   
4                0.0  org.apache:cayenne  ...               0   

  confirmed_issues  sqale_debt_ratio  code_smells  bugs  reliability_rating  \
0                0               3.5        11635  1438                   5   
1                0               3.5        11635  1438                   5   
2                0               3.5        11635  1438                   5   
3                0               3.5        11635  1438                   5   
4                0               3.5        11635  1438                   5   

   vulnerabilities  security_rating   files  comment_lines_density  
0              112                4  2354.0                   11.0  
1              112                4  2354.0                   11.0  
2              112                4  2354.0                   11.0  
3              112                4  2354.0                   11.0  
4              112                4  2354.0                   11.0  

[5 rows x 36 columns]

In [99]:
sonar_complete.shape[0]

1197656

In [100]:
# deleting duplicated rows
sonar_complete = sonar_complete[sonar_complete.duplicated()]
sonar_complete.shape[0]

1083655

FALTA AGRUPAR LES VARIBLES NUMERIQUES COM VULGUEM (AL FICAR-HO PER AUTOR FE MEAN O SUM DEL QUE VULGUEM COM PER EXEMPLE DEL ISSUE_CODE_LENGTH)

In [103]:
s = sonar_complete['analysis_key'].unique()
len(s)

9594